core/engine/src/value/conversions/try_from_js/collections.rs

```diff
@@ -62,3 +95,44 @@ where
             .collect()
     }
 }
+
+fn for_each_elem_in_js_map<F>(js_map: &JsMap, mut f: F, context: &mut Context) -> JsResult<()>
+where
+    F: FnMut(JsValue, JsValue, &mut Context) -> JsResult<()>,
+{
+    let unexp_obj_err = || {
+        JsResult::Err(
+            JsNativeError::typ()
+                .with_message("MapIterator return unexpected object")
+                .into(),
+        )
+    };
```

Comment:
On second thought, this looks like something that should replace `JsMap::for_each`, right? I think it would be a lot more ergonomic instead of our current API.

Reply:
~Yes, but I actually not sure how to do it via `JsMp::for_each` because you should to create an `JsFunction` and how I can see it can be done via `NativeFunctionPointer` but you can't capture in such function outer `map` Rust object (`HashMap`/`BTreeMap`). And if you use closure that will capture the `map`, then it seems like you can't create `JsFunction`.

Actually we don't want to create extra js objects here (like `JsFunction`), and therefor maybe it makes sense to create an function in `JsMap` that will do `for_each` but for Rust fn (for `F: FnMut`), and maybe make it public only in the crate.

Reply:
Yes, I was more talking about substituting `JsMap::for_each` with this new function, since it seems that passing `JsFunction` directly is a lot more painful overall.

Reply:
Done

Is `riust_for_each` acceptable name?

Just for sureness, is [this](https://github.com/Nikita-str/boa/blob/try-from-js--JsMap/core/engine/src/builtins/map/mod.rs#L479-L501) equivalent to the next:
```rust
let map = map
    .downcast_ref::<OrderedMap<JsValue>>()
    .expect("checked that `this` was a map");

map.iter().try_for_each(|(k, v)|{
    let args = [v.clone(), k.clone(), this.clone()];
    callback.call(this_arg, &args, context).map(|_|())
})?;
Ok(JsValue::undefined())
```
If they are not the same, then seems like I incorrectly impl the `fn rust_for each`

If they the same, maybe simplify it?

Reply:
Not really sure, but I think we have pretty good coverage. You can try replacing it to see if there are any regressions on the test262 suite.

Reply:
No, they are not the same because `JsMap` can be changed in `for_each` and it will panic because of mut and non-mut borrowing at the same time